lib/cn/kvset_builder.c

```diff
@@ -339,30 +348,37 @@ kvset_builder_finish(struct kvset_builder *imp)
 {
     merr_t err = 0;
 
-    assert(imp->kbb);
-    assert(imp->vbb);
+    INVARIANT(imp->hbb);
+    INVARIANT(imp->kbb);
+    INVARIANT(imp->vbb);
 
-    err = kbb_finish(imp->kbb, &imp->kblk_list, imp->seqno_min, imp->seqno_max);
+    err = kbb_finish(imp->kbb, &imp->kblk_list);
     if (ev(err))
         return err;
 
     if (imp->kblk_list.n_blks == 0) {
-        /* There are no kblocks. This happens when each input key has
-         * a tombstone and we are in "drop_tomb" mode.  This
-         * output kvset is empty and should not be created.  Destroy
-         * the corresponding vblock generator (which aborts any
-         * mblocks it has already allocated) and move on.  The empty
-         * kblk_list will prevent htis kvset from being created. */
+        /* There are no kblocks. This happens when each input key has a
+         * tombstone and we are in "drop_tomb" mode. This output kvset is empty
+         * and should not be created. Destroy the corresponding hblock vblock
+         * builder (which aborts any mblocks they had already allocated) and
+         * move on. The empty kblk_list will prevent this kvset from being
+         * created.
+         */
         vbb_destroy(imp->vbb);
-        imp->vbb = 0;
-        return 0;
+        imp->vbb = NULL;
+    } else {
+        err = vbb_finish(imp->vbb, &imp->vblk_list);
+        if (ev(err))
+            return err;
     }
 
-    err = vbb_finish(imp->vbb, &imp->vblk_list);
+    /* A constructed kvset will always only have 1 vgroup */
+    err = hbb_finish(imp->hbb, &imp->hblk, imp->seqno_min, imp->seqno_max, imp->kblk_list.n_blks,
+        imp->vblk_list.n_blks, 1, kbb_get_composite_hlog(imp->kbb));
     if (ev(err))
```

Comment:
If hbb_finish() fails, where do we abort the kblocks and vblocks that are in the local lists: imp->kblk_list and imp->vblk_list?

Reply:
Good point Nabeel. It seems like in some places after kvset_builder_get_mblocks() is called, there is cleanup, but in other places not. It isn't consistent. I will add the necessary cleanup into this function.

Reply:
Please review and let me know if this is what you expect.

Reply:
Looks good to me, thanks.. Btw, if we fail to abort it here in the error path then mpool reclaims these uncommitted mblocks during the subsequent open.